### DISCUSSION TITLE
编译问题

### DISCUSSION THREAD

**Discussion by liuxufeng2016** [YanMing-lxb/GUET_Thesis_LaTeX#38] [提问区]
采用最新的模版，从以下链接下载：
https://github.com/GUET-TeX-Users-Group/Guet_Thesis_LaTeX.git
编译环境：vscode
编译提示3个错误
I found no \citation bibstyle
I found no \citation bibdata
I found no \citation commands
![Uploading 错误提示.png…]()

**Comment by liuxufeng2016** [comment]
感谢本模版的大力支持。问题已经解决了。但我不知道如何删除问题。算了自问自答了。
问题：需要biber编译，我之前用的了BibTex。
解决方法：需要配置编译链，即设置settings.json。
具体方法参照如下链接：
https://blog.csdn.net/weixin_53011291/article/details/139034367
可能理解不是很专业，但能解决问题就行。对于非专业为员，能解决问题就好，可能也不需要知道为什么他是这样的。哈哈。扯远了。

**Reply by tangfen12** [reply]
同样的问题，用这个方法解决了。不过我在用完这个方法之后，错误提示虽然消失了，但是没有pdf输出，而且左下角一直是红叉。我将文件全部删除，重新下载，换了个文件名，重新打开。这样就好了，可以正常输出了。

**Reply by YanMing-lxb** [reply]
解决这个问题，还是编译链的问题，推荐使用latexmk 或者我自己开发的pytexmk ，关于vscode的配置和latexmk的配置在我写的教程中有详细说明 [YM-VSCode-Configurations-for-LaTeX](https://github.com/YanMing-lxb/YM-VSCode-Configurations-for-LaTeX)，可以看一下。
也可以使用[PyTeXMK](https://github.com/YanMing-lxb/PyTeXMK)，详细说明见github仓库

**Comment by liuxufeng2016** [comment]
您好，非常感谢，上次配置好环境后，又出现编译不通过了。日志文件看了半天也没看懂。搞不请具体问题是什么？辛苦有空帮瞅瞅。谢谢。以下是日志文件和界面截图。
[main.log](https://github.com/user-attachments/files/18240059/main.log)
![1](https://github.com/user-attachments/assets/034a8482-1cdf-450e-88d4-ba89fc51c672)

**Reply by YanMing-lxb** [reply]
main.tex:106: Undefined control sequence.
<argument> Chapters/ChThesisBibResourceapter 
                                             5
l.106 ...nput{Chapters/Ch\ThesisBibResourceapter5}
                                           

看日志文件最后面
106 行有个没定义的命令，路径应该用/不要用\

**Comment by liuxufeng2016** [comment]
非常感谢，找到问题原因了，问题已经解决。
报错的代码：input{Chapters/Ch\ThesisBibResourceapter5}
正确代码是：\input{Chapters/Chapter5}&nbsp;&nbsp;
应该是有一次粘贴时候，光标定位了错误的位置，误粘贴了“\ThesisBibResource”


此外，如果编译出错了，错误原因都是在main.log最后部分是吧





刘旭锋
***@***.***



&nbsp;






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "YanMing-lxb/GUET_Thesis_LaTeX"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年12月24日(星期二) 晚上11:23
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [YanMing-lxb/GUET_Thesis_LaTeX] 编译问题 (Discussion #38)





 
main.tex:106: Undefined control sequence.
  Chapters/ChThesisBibResourceapter
 5
 l.106 ...nput{Chapters/Ch\ThesisBibResourceapter5}
 
看日志文件最后面
 106 行有个没定义的命令，路径应该用/不要用\

**Reply by YanMing-lxb** [reply]
一般都是在后面，一下看还是能找到的，实在不行可以问大模型，一般都能解决

**Comment by liuxufeng2016** [comment]
认真阅读您写的README，我自己也试着在另外一台电脑上按装了一遍，不出意外还是没有成功。
个人理解要在VS Code编辑器上使用PyTeXMK编译器运行TeX，需要遵循以下步骤：
首先，确保安装了**TeXLive和VS Code编辑器**这两个关键软件。
接下来，进行VS Code环境的配置。这包括几个具体步骤：**安装LaTeX Workshop插件，配置setting.json文件，以及安装pytexmk编译器**。在配置setting.json文件时，需要安装一个PDF查看软件（例如smartPDF），并在该文件中正确设置VS Code和smartPDF的启动路径。
此外，安装PyTeXMK编译器之前，必须先确保安装了Python 3.8或更高版本。完成Python的安装后，打开cmd命令行，输入pip install pytexmk命令来安装pytexmk。
我按照上述思路完成了所有工作，但编译结果提示无法识别：pytexmk，错误提示如下图所示。
![界面](https://github.com/user-attachments/assets/c5184e0c-0a59-48dd-8d5a-9781c623276a)
以上是已经完成的步骤结果：
1、setting.json已经按照网页中给的配置代码设置：
网页链接:[[YM-VSCode-Configurations-for-LaTeX/Docs/LaTeX-Workshop配置说明.md at main · YanMing-lxb/YM-VSCode-Configurations-for-LaTeX](https://github.com/YanMing-lxb/YM-VSCode-Configurations-for-LaTeX/blob/main/Docs/LaTeX-Workshop%E9%85%8D%E7%BD%AE%E8%AF%B4%E6%98%8E.md)](url)
配置代码放见如下settingjson.txt文件
[settingjson.txt](https://github.com/user-attachments/files/18305326/settingjson.txt)
2、Python 3.8已经安装，见下图
![ptyone](https://github.com/user-attachments/assets/540f77b2-3487-4229-bdbf-838d9f6962b0)
3、Pytexmk已经按装，见下图
![pytexmk](https://github.com/user-attachments/assets/0b3cd7ac-c876-4380-bfbc-9d3a6a45270e)

**Reply by YanMing-lxb** [reply]
这个是这样的，你把下图里的终端从powershell 换成 CMD 终端。win 里有两个终端，要用CMD终端
![image](https://github.com/user-attachments/assets/e6ca05d9-fc70-418c-bb95-e084742300ba)

**Comment by YanMing-lxb** [comment]
不对，是我说错了，一般用cmd终端会成功的，其实powershell也能成功，不过你这个powershell出问题，可能是python没有接入powershell,，这个具体操作我就不会了。

**Comment by liuxufeng2016** [comment]
非常感谢，参考您的提示，已经找到原因并解决了。是因为tex文件名不是main，修改了文件名就可以了。




刘旭锋
***@***.***



&nbsp;






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "YanMing-lxb/GUET_Thesis_LaTeX"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年1月4日(星期六) 中午12:22
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [YanMing-lxb/GUET_Thesis_LaTeX] 编译问题 (Discussion #38)





 
不对，是我说错了，一般用cmd终端会成功的，其实powershell也能成功，不过你这个powershell出问题，可能是python没有接入powershell,，这个具体操作我就不会了。

**Comment by liuxufeng2016** [comment]
我修改了Accomplishs.bib，编译的PDF成果没有还是原模版的。尤其是只保留了一项专利和1篇论文，删除了其余成果，结果很是原模版的效果。
ThesisAchResource目录没有修改，如图：
![界面](https://github.com/user-attachments/assets/5a3bfcf5-1b66-40fe-bef1-ed51e1bfe530)
Accomplishs.bib，文件
![bib](https://github.com/user-attachments/assets/2a718821-6247-4ada-8512-b7a6187a190e)
pdf效果
![2](https://github.com/user-attachments/assets/5e475376-76fc-4899-9764-afa1a60b2fa3)

**Reply by YanMing-lxb** [reply]
把辅助文件清除一下，然后重新编译试试， `pytexmk -C`

**Comment by liuxufeng2016** [comment]
还真是，谢谢，非常感谢



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2025年1月6日(周一) 晚上11:08
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [YanMing-lxb/GUET_Thesis_LaTeX] 编译问题 (Discussion #38)




 
把辅助文件清除一下，然后重新编译试试， pytexmk -C

**Comment by liuxufeng2016** [comment]
尊敬的YanMing-lxb师兄：

您好！

首先，非常感谢您开发了pytexmk这款编辑器，它极大地提升了我撰写英文论文的效率，特别是在与VS Code配合使用时，我几乎已经完全放弃了TeXstudio。

不过，在使用pytexmk的过程中，我遇到了一个小问题，希望能得到您的解答。我注意到，在配置和使用pytexmk时，似乎要求主文件名必须是main.tex。我原本的理解是，文件名只要符合由英文字母和数字组成，并以英文字母开头的规则即可。然而，实际操作中似乎只有命名为main.tex的文件才能被正确识别和处理。

这一限制在实际办公场景中给我带来了一些不便，因为有时候我需要同时处理多个项目或文档，每个项目或文档都可能有自己的主文件，而这些主文件的名字往往根据内容或项目名称来命名，而非统一为main.tex。

因此，我想请问：

pytexmk是否确实要求主文件名必须是main.tex？
如果有这样的限制，是否有可能在未来的版本中提供对自定义主文件名的支持？
如果目前不支持自定义主文件名，是否有其他方法或配置可以绕过这一限制，以便我能够更灵活地管理我的文档和项目？
再次感谢您的辛勤工作和无私奉献，期待您的回复！

祝好，
XF L

**Reply by YanMing-lxb** [reply]
感谢您的赞美，pytexmk 相关的内容可以到 [pytexmk 仓库](https://github.com/YanMing-lxb/PyTeXMK) 进行提问。

你看一下 [待编译主文件及编译类型选定逻辑](https://github.com/YanMing-lxb/PyTeXMK?tab=readme-ov-file#%E5%BE%85%E7%BC%96%E8%AF%91%E4%B8%BB%E6%96%87%E4%BB%B6%E5%8F%8A%E7%BC%96%E8%AF%91%E7%B1%BB%E5%9E%8B%E9%80%89%E5%AE%9A%E9%80%BB%E8%BE%91) 这里的说明，看看能不能解决你的疑问。

- 如果有看不懂的地方欢迎告诉我，这块的说明我可能写的不太容易懂。
- 如果有新的需求也请告诉我。不过建议在 pytexmk 的仓库中提出。

**Reply by liuxufeng2016** [reply]
非常抱歉，由于这个网页环境我还不熟悉，导致在不适宜的仓库中提问了问题。下次会注意的。不过既然已经在这里提问了，我就在此作个回复，可能有点长。
我反复阅读了文档说明，尽管内容很专业，创作者可能很容易理解，甚至可能在写作时是清晰的。然而，对于万千读者来说，确实有些难以理解，这或许是“类程序员”的一个通病。不再吐槽了，以下是作为新手的我，对说明文档的理解、我遇到的问题描述，及解决过程。
**1. 说明文档**
[[YanMing-lxb/PyTeXMK: LaTeX 辅助编译命令行程序 LaTeX Auxiliary Compilation Command Line Tool](https://github.com/YanMing-lxb/PyTeXMK?tab=readme-ov-file#%E9%85%8D%E7%BD%AE%E6%96%87%E4%BB%B6%E8%AF%B4%E6%98%8E)](url)
网页 中提到“2.待编译主文件名：main.tex”，见下图，意味着默认编译的文件名是“main.tex”。

单击“配置文件说明”，其链接内容就在本页面的下半部分：
![2](https://github.com/user-attachments/assets/67dbec4e-1dab-4f4b-9884-3c0036351d63)

 
解释：
第1 点：如果命令行参数中指定了主文件，则使用该主文件名。
**（这里的命令参数中指定了主文件是要在的pytexmk后面加入主文件名吗？比如 pytexmk liu0112.tex，是这样吗？ ）**
第3点：如果存在魔法注释 % !TEX root，则根据魔法注释指定的文件作为主文件。
**（魔法注释 % !TEX root，应该是在tex中加入这个命令，前200行），这个我没有尝试，因为目前还用不到。**
**2. 问题描述**
我撰写的是论文回复信，有多个审稿人，每个审稿人一个tex文档，如下图，Reviewer1.tex、Reviewer2.tex、Reviewer3.tex、Reviewer4.tex、Reviewer5.tex。当然还包括了自认为的主文件“review-response.tex”。因为多个tex文件在一个文件件下，且没有指定main.tex，当然此模版也没有用魔法注释，所以计算机无法判断哪个是主文件。
大胆的猜测下，只要多个子tex文件放在了文件夹中，不和所谓的主文件放同一个文件夹中，或只要主文件用main命令，或只要在主文件中加入魔法注释 % !TEX root，计算机应该能判断哪个是主文件。我没有尝试，仅仅个人揣测而已。
 
![3](https://github.com/user-attachments/assets/9743d595-419b-44ca-8bea-ddaf6e93f2ac)

**3. 解决过程**
将5个审稿人对应的tex放在子文件夹中，主文件名改为main，如下图。采用编译命令pytexmk即可编译。
 
![4](https://github.com/user-attachments/assets/dc6d6a0d-2871-4866-8a00-cd5dbc50d698)


当然将5个审稿人对应的tex放在子文件夹中后，主文件名改为其他名字也是可以的，比如Response01121.txt，可以英文字母和数字组合。这样用的好处在于不同工程文件用不用主文名，避免统一用main命令带来阅读上的障碍。

**Reply by YanMing-lxb** [reply]
其实你这就属于`待编译主文件选定逻辑`的第四条，根目录下存在多个tex文件，但是只有一个tex文件中存在 `\documentclass[]{}`或 `\begin{document}` 命令（这个文件就是主文件）。所以只在终端输入 pytexmk 就行

**Reply by YanMing-lxb** [reply]
整体判断逻辑是这样的，当 输入 pytexmk 主文件名 时，优先使用终端中指定的主文件名
如果项目根目录下只有一个 tex 文件，则默认这个 tex 文件就是主文件
如果存在多个主文件（就是具有\begin{document}这个命令的tex文件）同时终端也没有指定要被编译的主文件，则优先编译具有魔法注释 root 指定的tex文件
如果存在多个tex文件单只有一个主文件，没有终端指定待编译主文件，同时也没有魔法注释指定主文件，那么就检索根目录下的所有tex文件，存在`\documentclass[]{}`或 `\begin{document}` 命令的tex文件就是待编译主文件了。为了防止出现tex文件过大而导致检索耗时过长的情况，我就加了个限制，只检索前200行
如果上述情况都不存在，那就尝试将main作为待编译主文件，看看能不能编译
再不行就报错了。

你看看，这样的话，能否理解我的意图。

**Comment by liuxufeng2016** [comment]
非常感谢您的解答，我理解您的意思了，谢谢。


刘旭锋
***@***.***



&nbsp;






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "YanMing-lxb/GUET_Thesis_LaTeX"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年1月12日(星期天) 晚上9:35
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [YanMing-lxb/GUET_Thesis_LaTeX] 编译问题 (Discussion #38)





 
整体判断逻辑是这样的，当 输入 pytexmk 主文件名 时，优先使用终端中指定的主文件名
 如果项目根目录下只有一个 tex 文件，则默认这个 tex 文件就是主文件
 如果存在多个主文件（就是具有\begin{document}这个命令的tex文件）同时终端也没有指定要被编译的主文件，则优先编译具有魔法注释 root 指定的tex文件
 如果存在多个tex文件单只有一个主文件，没有终端指定待编译主文件，同时也没有魔法注释指定主文件，那么就检索根目录下的所有tex文件，存在\documentclass[]{}或 \begin{document} 命令的tex文件就是待编译主文件了。为了防止出现tex文件过大而导致检索耗时过长的情况，我就加了个限制，只检索前200行
 如果上述情况都不存在，那就尝试将main作为待编译主文件，看看能不能编译
 再不行就报错了。
 
你看看，这样的话，能否理解我的意图。

**Comment by liuxufeng2016** [comment]
你好，你的来信已收到，我将尽快给你回复您。谢谢。Hello.Your letter has been received. I will reply to you as soon as possible.Thank you.This is an automatic mail reply.Your message has been delivered successfully。

**Comment by chendongmei11111** [comment]
尊敬的师兄您好！

我在使用 GUET-Thesis 模板时遇到一个问题，希望能得到您的帮助。

在参考文献中使用 @online 类型时，URL 和访问日期（urldate）无法正常显示，仅显示 [EB]（这里应该显示 [EB/OL]），但链接和日期未输出。.bib 文件中已正确填写 url 和 urldate 字段，例如：
@online{MPS_MotorVehicle2025,  % 推荐用@online，比@misc更贴合网络文献
	author        = {中华人民共和国公安部},
	title         = {全国机动车保有量达4.53亿辆 驾驶人达5.42亿人},
	howpublished  = {[EB/OL]},  % 若gbmedium=true，此句可省略（模板会自动加载体标识）
	year          = {2025},
	month         = {01},
	day           = {18},
	url           = {https://www.mps.gov.cn:9080/n2254314/n6409334/c9939035/content.html},
	urldate       = {2025-09-12}  % 格式需为“年-月-日”，与urldate=long匹配
}
请问是否需要特殊设置或修改样式文件才能使 URL 和访问日期正确显示？感谢您抽出时间查看，祝您生活愉快，工作顺利！

此致

敬礼！
